### PR TITLE
Add structured trace profiler

### PR DESCRIPTION
**Pull request by Kha**
The `--profile` output is nice for getting a coarse summary of what time is spent on within a file or package, but not as helpful for finding specific locations, especially since the profiler categories are disjoint (so a tactic invocation spending most of the time in many small typeclass synthesis calls will never show up in the profiler). The timing information in the structured traces is better for this use case, but you need to know what trace class to activate in the first place and you likely will still get lost in its pages of output.

This PR introduces a new option `trace.profiler` (and `trace.profiler.threshold` defaulting to 10ms) that is independent of the existing profiler option and aims to make finding and exploring expensive locations in traces easier by activating trace nodes not by class but by whether their (inclusive!) execution time exceeds the configured threshold.
```shell-session
$ lean Mathlib/Data/Polynomial/FieldDivision.lean -Dtrace.profiler=true -Dpp.oneline=true
[Elab.input] [5.410904s] theorem derivative_rootMultiplicity_of_root [CharZero R] {p : R[X]} {t : R} (hpt : p.IsRoot t) : [...]
  [Elab.step] [5.312021s] rcases eq_or_ne p 0 with (rfl | hp) [...]
    [Elab.step] [5.311969s] rcases eq_or_ne p 0 with (rfl | hp) [...]
      [Elab.step] [0.065163s] · simp
        [Elab.step] [0.065141s] simp
          [Elab.step] [0.065134s] simp
            [Meta.isDefEq] [0.023725s] ❌ ↑?f (OfNat.ofNat ?n) =?= ↑Polynomial.derivative 0
              [Meta.isDefEq] [0.023450s] ❌ OfNat.ofNat ?n =?= 0
                [Meta.isDefEq] [0.023335s] ❌ instOfNat =?= Zero.toOfNat0
                  [Meta.isDefEq] [0.023310s] ❌ { ofNat := ↑0 } =?= { ofNat := Zero.zero }
                    [Meta.isDefEq] [0.023263s] ❌ ↑0 =?= Zero.zero
                      [Meta.isDefEq] [0.023224s] ❌ NatCast.natCast 0 =?= Zero.zero
                        [Meta.isDefEq] [0.023184s] ❌ NonAssocSemiring.toNatCast.1 0 =?= Polynomial.zero.1
                          [Meta.isDefEq] [0.023157s] ❌ NonAssocSemiring.toNatCast.1 0 =?= { toFinsupp := 0 }
                            [Meta.isDefEq.onFailure] [0.013963s] ❌ NonAssocSemiring.toNatCast.1 0 =?= { toFinsupp := 0 }
                              [Meta.isDefEq] [0.013937s] ❌ NonAssocSemiring.toNatCast.1 0 =?= { toFinsupp := 0 }
                                [Meta.isDefEq] [0.013796s] ❌ ↑0 =?= { toFinsupp := 0 }
            [Meta.isDefEq] [0.017557s] ✅ ↑?f 0 =?= ↑Polynomial.derivative 0
              [Meta.synthInstance] [0.016535s] ✅ ZeroHomClass (R[X] →ₗ[R] R[X]) R[X] R[X]
            [Meta.synthInstance] [0.014352s] ✅ ZeroHomClass (R[X] →ₗ[R] R[X]) R[X] R[X]
      [Elab.step] [0.131247s] simp only [derivative_pow, derivative_mul, derivative_sub, derivative_X, derivative_C, sub_zero, mul_one]
        [Meta.isDefEq] [0.048698s] ✅ ↑Polynomial.derivative (?p ^ ?n) =?= ↑Polynomial.derivative ((Polynomial.X - ↑Polynomial.C t) ^ Polynomial.rootMultiplicity t p)
          [Meta.isDefEq] [0.019443s] ✅ R[X] →ₗ[R] R[X] =?= R[X] →ₗ[R] R[X]
            [Meta.isDefEq] [0.016730s] ✅ NonUnitalNonAssocSemiring.toAddCommMonoid =?= NonUnitalNonAssocSemiring.toAddCommMonoid
              [Meta.isDefEq] [0.016721s] ✅ NonAssocSemiring.toNonUnitalNonAssocSemiring.1 =?= NonAssocSemiring.toNonUnitalNonAssocSemiring.1
                [Meta.isDefEq] [0.016591s] ✅ NonUnitalNonAssocSemiring.toAddCommMonoid =?= NonUnitalNonAssocSemiring.toAddCommMonoid
                  [Meta.isDefEq] [0.016581s] ✅ (Function.Injective.nonUnitalNonAssocSemiring Polynomial.toFinsupp (_ : Function.Injective Polynomial.toFinsupp) [...] =?= (Function.Injective.nonUnitalNonAssocSemiring Polynomial.toFinsupp (_ : Function.Injective Polynomial.toFinsupp) [...]
                    [Meta.isDefEq] [0.016527s] ✅ AddCommMonoid.mk (_ : ∀ (a b : R[X]), a + b = b + a) =?= AddCommMonoid.mk (_ : ∀ (a b : R[X]), a + b = b + a)
                      [Meta.isDefEq] [0.015472s] ✅ AddCommMonoid.toAddMonoid =?= AddCommMonoid.toAddMonoid
                        [Meta.isDefEq] [0.015461s] ✅ (Function.Injective.addCommMonoid Polynomial.toFinsupp (_ : Function.Injective Polynomial.toFinsupp) [...] =?= (Function.Injective.addCommMonoid Polynomial.toFinsupp (_ : Function.Injective Polynomial.toFinsupp) [...]
                          [Meta.isDefEq] [0.015388s] ✅ AddMonoid.mk (_ : ∀ (a : R[X]), 0 + a = a) (_ : ∀ (a : R[X]), a + 0 = a) AddMonoid.nsmul =?= AddMonoid.mk (_ : ∀ (a : R[X]), 0 + a = a) (_ : ∀ (a : R[X]), a + 0 = a) AddMonoid.nsmul
          [Meta.isDefEq] [0.017521s] ✅ LinearMap.instFunLikeLinearMap =?= LinearMap.instFunLikeLinearMap
            [Meta.isDefEq.delta] [0.017476s] ✅ LinearMap.instFunLikeLinearMap =?= LinearMap.instFunLikeLinearMap
              [Meta.isDefEq] [0.015240s] ✅ NonUnitalNonAssocSemiring.toAddCommMonoid =?= NonUnitalNonAssocSemiring.toAddCommMonoid
                [Meta.isDefEq] [0.015231s] ✅ NonAssocSemiring.toNonUnitalNonAssocSemiring.1 =?= NonAssocSemiring.toNonUnitalNonAssocSemiring.1
                  [Meta.isDefEq] [0.015127s] ✅ NonUnitalNonAssocSemiring.toAddCommMonoid =?= NonUnitalNonAssocSemiring.toAddCommMonoid
                    [Meta.isDefEq] [0.015118s] ✅ (Function.Injective.nonUnitalNonAssocSemiring Polynomial.toFinsupp (_ : Function.Injective Polynomial.toFinsupp) [...] =?= (Function.Injective.nonUnitalNonAssocSemiring Polynomial.toFinsupp (_ : Function.Injective Polynomial.toFinsupp) [...]
                      [Meta.isDefEq] [0.015092s] ✅ AddCommMonoid.mk (_ : ∀ (a b : R[X]), a + b = b + a) =?= AddCommMonoid.mk (_ : ∀ (a b : R[X]), a + b = b + a)
                        [Meta.isDefEq] [0.014201s] ✅ AddCommMonoid.toAddMonoid =?= AddCommMonoid.toAddMonoid
                          [Meta.isDefEq] [0.014191s] ✅ (Function.Injective.addCommMonoid Polynomial.toFinsupp (_ : Function.Injective Polynomial.toFinsupp) [...] =?= (Function.Injective.addCommMonoid Polynomial.toFinsupp (_ : Function.Injective Polynomial.toFinsupp) [...]
                            [Meta.isDefEq] [0.014165s] ✅ AddMonoid.mk (_ : ∀ (a : R[X]), 0 + a = a) (_ : ∀ (a : R[X]), a + 0 = a) AddMonoid.nsmul =?= AddMonoid.mk (_ : ∀ (a : R[X]), 0 + a = a) (_ : ∀ (a : R[X]), a + 0 = a) AddMonoid.nsmul
        [Meta.isDefEq] [0.036693s] ✅ ↑Polynomial.derivative (?f - ?g) =?= ↑Polynomial.derivative (Polynomial.X - ↑Polynomial.C t)
          [Meta.isDefEq] [0.018497s] ✅ R[X] →ₗ[R] R[X] =?= R[X] →ₗ[R] R[X]
            [Meta.isDefEq] [0.015961s] ✅ NonUnitalNonAssocSemiring.toAddCommMonoid =?= NonUnitalNonAssocSemiring.toAddCommMonoid
              [Meta.isDefEq] [0.015952s] ✅ NonAssocSemiring.toNonUnitalNonAssocSemiring.1 =?= NonAssocSemiring.toNonUnitalNonAssocSemiring.1
                [Meta.isDefEq] [0.015822s] ✅ NonUnitalNonAssocSemiring.toAddCommMonoid =?= NonUnitalNonAssocSemiring.toAddCommMonoid
                  [Meta.isDefEq] [0.015812s] ✅ (Function.Injective.nonUnitalNonAssocSemiring Polynomial.toFinsupp (_ : Function.Injective Polynomial.toFinsupp) [...] =?= (Function.Injective.nonUnitalNonAssocSemiring Polynomial.toFinsupp (_ : Function.Injective Polynomial.toFinsupp) [...]
                    [Meta.isDefEq] [0.015762s] ✅ AddCommMonoid.mk (_ : ∀ (a b : R[X]), a + b = b + a) =?= AddCommMonoid.mk (_ : ∀ (a b : R[X]), a + b = b + a)
                      [Meta.isDefEq] [0.014808s] ✅ AddCommMonoid.toAddMonoid =?= AddCommMonoid.toAddMonoid
                        [Meta.isDefEq] [0.014798s] ✅ (Function.Injective.addCommMonoid Polynomial.toFinsupp (_ : Function.Injective Polynomial.toFinsupp) [...] =?= (Function.Injective.addCommMonoid Polynomial.toFinsupp (_ : Function.Injective Polynomial.toFinsupp) [...]
                          [Meta.isDefEq] [0.014761s] ✅ AddMonoid.mk (_ : ∀ (a : R[X]), 0 + a = a) (_ : ∀ (a : R[X]), a + 0 = a) AddMonoid.nsmul =?= AddMonoid.mk (_ : ∀ (a : R[X]), 0 + a = a) (_ : ∀ (a : R[X]), a + 0 = a) AddMonoid.nsmul
          [Meta.isDefEq] [0.017675s] ✅ LinearMap.instFunLikeLinearMap =?= LinearMap.instFunLikeLinearMap
            [Meta.isDefEq.delta] [0.017629s] ✅ LinearMap.instFunLikeLinearMap =?= LinearMap.instFunLikeLinearMap
              [Meta.isDefEq] [0.015387s] ✅ NonUnitalNonAssocSemiring.toAddCommMonoid =?= NonUnitalNonAssocSemiring.toAddCommMonoid
                [Meta.isDefEq] [0.015377s] ✅ NonAssocSemiring.toNonUnitalNonAssocSemiring.1 =?= NonAssocSemiring.toNonUnitalNonAssocSemiring.1
                  [Meta.isDefEq] [0.015273s] ✅ NonUnitalNonAssocSemiring.toAddCommMonoid =?= NonUnitalNonAssocSemiring.toAddCommMonoid
                    [Meta.isDefEq] [0.015264s] ✅ (Function.Injective.nonUnitalNonAssocSemiring Polynomial.toFinsupp (_ : Function.Injective Polynomial.toFinsupp) [...] =?= (Function.Injective.nonUnitalNonAssocSemiring Polynomial.toFinsupp (_ : Function.Injective Polynomial.toFinsupp) [...]
                      [Meta.isDefEq] [0.015239s] ✅ AddCommMonoid.mk (_ : ∀ (a b : R[X]), a + b = b + a) =?= AddCommMonoid.mk (_ : ∀ (a b : R[X]), a + b = b + a)
                        [Meta.isDefEq] [0.014329s] ✅ AddCommMonoid.toAddMonoid =?= AddCommMonoid.toAddMonoid
                          [Meta.isDefEq] [0.014319s] ✅ (Function.Injective.addCommMonoid Polynomial.toFinsupp (_ : Function.Injective Polynomial.toFinsupp) [...] =?= (Function.Injective.addCommMonoid Polynomial.toFinsupp (_ : Function.Injective Polynomial.toFinsupp) [...]
                            [Meta.isDefEq] [0.014292s] ✅ AddMonoid.mk (_ : ∀ (a : R[X]), 0 + a = a) (_ : ∀ (a : R[X]), a + 0 = a) AddMonoid.nsmul =?= AddMonoid.mk (_ : ∀ (a : R[X]), 0 + a = a) (_ : ∀ (a : R[X]), a + 0 = a) AddMonoid.nsmul
        [Meta.isDefEq] [0.020091s] ✅ ?a - 0 =?= 1 - 0
          [Meta.isDefEq] [0.019314s] ✅ 0 =?= 0
            [Meta.isDefEq] [0.019267s] ✅ Zero.toOfNat0 =?= Zero.toOfNat0
              [Meta.isDefEq.delta] [0.019253s] ✅ Zero.toOfNat0 =?= Zero.toOfNat0
                [Meta.isDefEq] [0.019221s] ✅ NegZeroClass.toZero =?= Polynomial.zero
                  [Meta.isDefEq] [0.019177s] ✅ NegZeroClass.toZero =?= { zero := { toFinsupp := 0 } }
                    [Meta.isDefEq] [0.019155s] ✅ Zero.zero =?= { toFinsupp := 0 }
                      [Meta.isDefEq] [0.019135s] ✅ Zero.zero.toFinsupp =?= 0
                        [Meta.isDefEq] [0.018959s] ✅ Zero.zero.1 =?= 0
                          [Meta.isDefEq] [0.018903s] ✅ Zero.zero.1 =?= Zero.toOfNat0.1
                            [Meta.isDefEq] [0.018875s] ✅ Zero.zero.1 =?= Zero.zero
                              [Meta.isDefEq] [0.018829s] ✅ Zero.zero.1 =?= MonoidWithZero.toZero.1
                                [Meta.isDefEq] [0.018763s] ✅ Zero.zero.1 =?= 0
                                  [Meta.isDefEq] [0.018714s] ✅ Zero.zero.1 =?= Zero.toOfNat0.1
                                    [Meta.isDefEq] [0.018685s] ✅ Zero.zero.1 =?= Zero.zero
                                      [Meta.isDefEq] [0.018638s] ✅ Zero.zero.1 =?= MulZeroClass.toZero.1
                                        [Meta.isDefEq] [0.018586s] ✅ Zero.zero.1 =?= 0
                                          [Meta.isDefEq] [0.018538s] ✅ Zero.zero.1 =?= Zero.toOfNat0.1
                                            [Meta.isDefEq] [0.018511s] ✅ Zero.zero.1 =?= Zero.zero
                                              [Meta.isDefEq] [0.018465s] ✅ Zero.zero.1 =?= AddMonoid.toZero.1
                                                [Meta.isDefEq] [0.018393s] ✅ Zero.zero.1 =?= Zero.zero
                                                  [Meta.isDefEq] [0.018344s] ✅ Zero.zero.1 =?= Finsupp.zero.1
                                                    [Meta.isDefEq] [0.018314s] ✅ Zero.zero.1 =?= { support := ∅, toFun := 0, mem_support_toFun := (_ : ∀ (x : ℕ), x ∈ ∅ ↔ OfNat.ofNat 0 x ≠ 0) }
                                                      [Meta.isDefEq] [0.016660s] ✅ Zero.zero.1.support =?= ∅
                                                        [Meta.isDefEq] [0.016634s] ✅ Zero.zero.1.1 =?= ∅
                                                          [Meta.isDefEq] [0.016585s] ✅ Zero.zero.1.1 =?= Finset.instEmptyCollectionFinset.1
                                                            [Meta.isDefEq] [0.016545s] ✅ Zero.zero.1.1 =?= Finset.empty
                                                              [Meta.isDefEq] [0.016500s] ✅ Zero.zero.1.1 =?= { val := 0, nodup := (_ : Multiset.Nodup 0) }
                                                                [Meta.isDefEq] [0.015981s] ✅ Zero.zero.1.1.val =?= 0
                                                                  [Meta.isDefEq] [0.015942s] ✅ Zero.zero.1.1.1 =?= 0
                                                                    [Meta.isDefEq] [0.015885s] ✅ Zero.zero.1.1.1 =?= Zero.toOfNat0.1
                                                                      [Meta.isDefEq] [0.015842s] ✅ Zero.zero.1.1.1 =?= Zero.zero
                                                                        [Meta.isDefEq] [0.015796s] ✅ Zero.zero.1.1.1 =?= Multiset.instZeroMultiset.1
                                                                          [Meta.isDefEq] [0.015753s] ✅ Zero.zero.1.1.1 =?= Multiset.zero
                                                                            [Meta.isDefEq] [0.015709s] ✅ Zero.zero.1.1.1 =?= ↑[]
                                                                              [Meta.isDefEq] [0.015664s] ✅ Zero.zero.1.1.1 =?= Quot.mk Setoid.r []
                                                                                [Meta.isDefEq.onFailure] [0.015593s] ✅ Zero.zero.1.1.1 =?= Quot.mk Setoid.r []
                                                                                  [Meta.synthInstance] [0.014838s] ✅ SubNegZeroMonoid R[X]
(...)
```
The new pretty printing option `pp.oneline` replaces anything but the first line with `[...]` to keep the output easily scannable. Note that because this option works on the pretty printer level, there may still be multi-line trace nodes, but the output seems reasonable enough.
```shell-session
                  [Elab.step] [0.018801s] refine if h : p.IsRoot t then ?pos✝ else ?neg✝
                    [Elab.step] [0.018756s] expected type: Polynomial.rootMultiplicity t p - 1 ≤ Polynomial.rootMultiplicity t (↑Polynomial.derivative p), term
                        if h : p.IsRoot t then ?pos✝ else ?neg✝
                      [Elab.step] [0.018747s] expected type: Polynomial.rootMultiplicity t p - 1 ≤ Polynomial.rootMultiplicity t (↑Polynomial.derivative p), term
                          let_mvar% ?m✝ := p.IsRoot t; wait_if_type_mvar% ?m✝; dite✝ ?m✝ (fun h => ?pos✝) (fun h => ?neg✝)
                        [Elab.step] [0.014474s] expected type: Polynomial.rootMultiplicity t p - 1 ≤ Polynomial.rootMultiplicity t (↑Polynomial.derivative p), term
                            wait_if_type_mvar% ?m✝; dite✝ ?m✝ (fun h => ?pos✝) (fun h => ?neg✝)
                          [Elab.step] [0.014434s] expected type: Polynomial.rootMultiplicity t p - 1 ≤ Polynomial.rootMultiplicity t (↑Polynomial.derivative p), term
                              dite✝ ?m✝ (fun h => ?pos✝) (fun h => ?neg✝)
                            [Meta.synthInstance] [0.014015s] ✅ Decidable (Polynomial.IsRoot p t)
```
`pp.oneline` currently produces non-interactive traces only.

Finally, as the output is otherwise regular trace output, we can in fact use @hargoniX's FlameTC explorer (after one [tiny generalization](https://github.com/Kha/FlameTC/commit/27969a35565b5de525a69338f1d5c19608b17da1)) to print nice-looking and even easier-to-explore flame graphs from it.
image

In the end, the output structure is not too different from what e.g. perf+hotspot can give us, but on a slightly more abstract, component-based level and importantly with more information about inputs and no danger of stacks getting cut down from too-deep recursion.

### CODE REVIEW

File: src/Lean/PrettyPrinter/Formatter.lean [hunk 1]
```diff
@@ -514,14 +514,28 @@ instance : Coe (Formatter → Formatter → Formatter) FormatterAliasValue := {
...
   trace[PrettyPrinter.format.input] "{Std.format stx}"
   let options ← getOptions
   let table ← Parser.builtinTokenTable.get
   catchInternalId backtrackExceptionId
     (do
-      let (_, st) ← (concat formatter { table := table, options := options }).run { stxTrav := Syntax.Traverser.fromSyntax stx };
-      pure $ Format.fill $ st.stack.get! 0)
+      let (_, st) ← (concat formatter { table, options }).run { stxTrav := .fromSyntax stx }
+      let mut f := st.stack[0]!
+      if (pp.oneline.get? options).getD (trace.profiler.get options) then
```
Comment (gebner) [inline src/Lean/PrettyPrinter/Formatter.lean:531]: Please don't couple `pp.oneline` to `trace.profiler`.  The `trace.profiler` option sounds super useful, I don't want it to cut off half the trace message and make them noninteractive.

File: src/Lean/Util/Trace.lean [hunk 2]
```diff
@@ -210,6 +236,12 @@ macro "trace[" id:ident "]" s:(interpolatedStr(term) <|> term) : doElem => do
     if (← Lean.isTracingEnabledFor cls) then
       Lean.addTrace cls $msg)
 
+macro "with_trace[" id:ident "]" s:(interpolatedStr(term) <|> termBeforeDo) " do " cont:doSeq : doElem => do
```
Comment (gebner) [inline src/Lean/Util/Trace.lean:239]: I'm not sure if this macro is worth it, it's used only three times in total and doesn't even support the fancier (and preferable) `withTraceNode*` combinators.

And unlike `trace[...]`, it computes `$msg` even if the trace class is disabled.

File: src/Lean/PrettyPrinter/Formatter.lean (hunk 1 above)
Comment (Kha) [reply src/Lean/PrettyPrinter/Formatter.lean:531]: I expected this to be the more useful default for my use cases, but I don't mind avoiding this option dependency

File: src/Lean/Util/Trace.lean (hunk 2 above)
Comment (Kha) [reply src/Lean/Util/Trace.lean:239]: Ah, a remnant of my original plan to simply upgrade the `trace` macro, which turned out not to be the best idea

**Comment by Kha** [comment]
Oh, I think I found the reason the diff doesn't load anymore...
image

**Comment by Kha** [comment]
!bench

**Comment by Kha** [comment]
!bench

**Comment by Kha** [comment]
Essentially no overhead anymore when not using the option, great. Since this change is purely additive and not a critical component, I think we can merge this and then refine later on if needed.

**Comment by Kha** [comment]
Interactive view of `Mathlib.Data.Polynomial.FieldDivision`, using @hargoniX' https://github.com/hargoniX/FlameTC/issues/1#issuecomment-1495933952:
https://www.speedscope.app/#profileURL=https://pp.ipd.kit.edu/~ullrich/tmp/flame.gz

**Pull request merged**